### PR TITLE
rpadmin: add AdminAddressesFromK8SDNS to rpadmin package

### PR DESCRIPTION
**Pull request by bojand**
Adds `AdminAddressesFromK8SDNS()` function which attempts to deduce admin API URLs based on [Kubernetes DNS resolution](https://github.com/kubernetes/dns/blob/master/docs/specification.md).

Also adds `UpdateAPIUrlsFromKubernetesDNS` which can be used to update the internal admin API URLs using `AdminAddressesFromK8SDNS()`.

### CODE REVIEW

File: rpadmin/admin.go [hunk 1]
```diff
@@ -630,3 +632,48 @@ func MaxRetries(r int) Opt {
...
+// AdminAddressesFromK8SDNS attempts to deduce admin API URLs
+// based on Kubernetes DNS resolution.
+// We are in Kubernetes.
+// https://github.com/kubernetes/dns/blob/master/docs/specification.md
+// Assume that Admin API URL configured is a Kubernetes Service URL.
+// This Admin API URL is passed in as the function argument.
+// Since it's a Kubernetes service, Kubernetes DNS creates a DNS SRV record
+// for the admin port mapping.
+// We can query the DNS record to get the target host names and ports.
+func AdminAddressesFromK8SDNS(adminAPIURL string) ([]string, error) {
```
Comment (weeco) [inline rpadmin/admin.go:645]: What do you think about making this a method of the `AdminAPI` rather than a standalone function? What I had in mind is that it uses the already configured URLs, and that you have a method such as 

```
func (a *AdminAPI) UpdateAPIUrlsFromKubernetesDNS(ctx context.context) error {}
```

and then it uses https://github.com/redpanda-data/common-go/blob/d034f64565442d1ca943efa0a420952db71a860d/rpadmin/admin.go#L84 as the initial URLs and will also mutate this set of URLs after successful discovery. Would that work and make sense @r-vasquez ?

My assumption is that this method has to be called oftentimes before issuing broker-specific adminapi requests, to ensure we have an up-to-date map and not only once when initializing the adminapi client.

File: rpadmin/admin.go [hunk 2]
```diff
@@ -630,3 +632,48 @@ func MaxRetries(r int) Opt {
...
+	}
+
+	_, records, err := net.LookupSRV("admin", "tcp", adminURL.Hostname())
+	if err != nil {
+		return nil, err
+	}
+
+	if len(records) == 0 {
+		return nil, nil
+	}
```
Comment (weeco) [inline rpadmin/admin.go:658]: I'm afraid we can run into nil pointer exceptions if not carefully checking the response. Should we return a checkable error in the case where we couldn't get any results. Do we end up here if the SRV record didn't exist as well or only if we got a response that doesn't map to any targets?

Comment (bojand) [reply rpadmin/admin.go]: I added `ErrNoSRVRecordsFound` error that is returned in case of no records... I am not really able to test this and mock it. I think this would be perhaps some weird edge case if possible at all... My intuition is that is the SRV records do not exist, we should get a [DNS error](https://pkg.go.dev/net#DNSError) since the query should return `NXDOMAIN`.

File: rpadmin/admin.go (hunk 1 above)
Comment (bojand) [reply rpadmin/admin.go:650]: Hmm interesting. I guess the workflow I had in mind was that the client (user of the API), would call `AdminAddressesFromK8SDNS` to get candidate URLs, and then call `NewHostClient()` to create a broker-specific Admin API Client. I am open to different approach.

**Comment by bojand** [comment]
#42 Seems somewhat related.

File: rpadmin/admin.go (hunk 1 above)
Comment (weeco) [reply rpadmin/admin.go:650]: I believe if this is becomes a more common usecase (besides the debug bundle work), this can quickly become cumbersome for the user for the admin api client. Therefore, I advocate for putting more logic inside this client, so that it becomes more handy to use.

**One option for that may be:**

Pros:
- Easy to use

Cons:
- Partial responses can become tricky in terms of error handling (see franz-go with sharded requests and error handling)

```go
  // Update internally stored URLs, so that we can talk to all brokers
  adminCl. UpdateAPIUrlsFromKubernetesDNS(ctx)

  // Get debug bundle status for one specific broker
  singleBrokerStatus, err := adminCl.GetDebugBundleStatus(brokerID)

  // Get debug bundle status for multiple brokers. Use nil to get status from all brokers
  statuses, err := adminCl.GetDebugBundleStatuses(brokerIDs)
```

**Another option:**

Pros:
- More generic approach for the adminapi client

Cons:
- You still have to take care of the logic when fetching from multiple / all brokers

```go
  // Update internally stored URLs, so that we can talk to all brokers
  adminCl. UpdateAPIUrlsFromKubernetesDNS(ctx)

  singleBrokerStatus, err := adminCl.ForBroker(id).GetDebugBundleStatus()
```

File: rpadmin/admin.go [hunk 3]
```diff
@@ -190,10 +197,22 @@ func newAdminAPI(urls []string, auth Auth, tlsConfig *tls.Config, dialer DialCon
...
+		return nil, err
+	}
+
+	return a, nil
+}
+
+func (a *AdminAPI) initURLs(urls []string, tlsConfig *tls.Config, forCloud bool) error {
+	if len(a.urls) != len(urls) {
+		a.urls = make([]string, len(urls))
+	}
```
Comment (r-vasquez) [inline rpadmin/admin.go:210]: Should this be instead if len(a.urls) == 0?

I wonder what is the expected result if len(a.urls) = 1, and len(urls) = 3... We will end up overwriting a.urls[0] and not sure if it's needed.

File: rpadmin/admin.go [hunk 4]
```diff
@@ -630,3 +649,63 @@ func MaxRetries(r int) Opt {
...
+	// redpanda-1.redpanda.redpanda.svc.cluster.local.
+	// take advantage of ordinals and order them accordingly
+	sort.Slice(records, func(i, j int) bool {
+		return records[i].Target < records[j].Target
+	})
+
+	urls := make([]string, 0, len(records))
+
+	proto := "http://"
+	if adminURL.Scheme == "https:" {
```
Comment (r-vasquez) [inline rpadmin/admin.go:687]: I think the parsed scheme from `url.Parse()` does not contain the colon
```suggestion
	if adminURL.Scheme == "https" {
```

File: rpadmin/admin_test.go [hunk 5]
```diff
@@ -222,3 +226,93 @@ func callsForNodeID(calls []testCall, nodeID int) []testCall {
 	}
 	return filtered
 }
+
+func TestAdminAddressesFromK8SDNS(t *testing.T) {
+	adminAPIURL := "http://redpanda-api.cluster.local:19644"
```
Comment (r-vasquez) [inline rpadmin/admin_test.go:231]: Can we add another test with https please?

File: rpadmin/admin_test.go [hunk 6]
```diff
@@ -222,3 +226,93 @@ func callsForNodeID(calls []testCall, nodeID int) []testCall {
...
+	assert.Equal(t, "http://rp-id123-2.rp-id123.redpanda.svc.cluster.local.:9644", brokerURLs[2])
+}
+
+func TestUpdateAPIUrlsFromKubernetesDNS(t *testing.T) {
+	adminAPIURL := "http://redpanda-api.cluster.local:19644"
+
+	adminAPIHostURL, err := url.Parse(adminAPIURL)
+	require.NoError(t, err)
+
+	srv, err := mockdns.NewServer(map[string]mockdns.Zone{
```
Comment (r-vasquez) [inline rpadmin/admin_test.go:278]: Very nice!

File: rpadmin/admin.go (hunk 3 above)
Comment (bojand) [reply rpadmin/admin.go:210]: `a.urls` could be initialized here, to original passed in urls to `NewClient()` which could be any number of elements.. This new code for `initURLs()` was moved from constructor to a helper function since it's used in two places now... It initializes _our_ `a.urls`... So we want to overwrite, which the loop will do... But in case the internal slice is not the correct size, we need to make it the correct size, so we check for equivalence, because the loop uses direct index `a.urls[i] = `.

Comment (r-vasquez) [reply rpadmin/admin.go:210]: I see, thanks!

File: rpadmin/admin_test.go (hunk 5 above)
Comment (bojand) [reply rpadmin/admin_test.go]: fixed

File: rpadmin/admin.go (hunk 4 above)
Comment (bojand) [reply rpadmin/admin.go]: fixed

File: rpadmin/admin.go (hunk 1 above)
Comment (bojand) [reply rpadmin/admin.go:667]: As per discussion, I have also added `UpdateAPIUrlsFromKubernetesDNS()` method to the Admin API struct.

File: rpadmin/admin.go [hunk 7]
```diff
@@ -190,20 +197,37 @@ func newAdminAPI(urls []string, auth Auth, tlsConfig *tls.Config, dialer DialCon
...
+
+const (
+	schemeHTTP  = "http"
+	schemeHTTPS = "https"
+)
+
+func (a *AdminAPI) initURLs(urls []string, tlsConfig *tls.Config, forCloud bool) error {
+	if len(a.urls) != len(urls) {
+		a.urls = make([]string, len(urls))
+	}
```
Comment (weeco) [inline rpadmin/admin.go:215]: Is the admin api supposed to be concurrency safe? If yes, we could have a race condition in the moment where we replace the URLs. Possibly we need to use a RWMutex and update the funcs that access `a.urls`

Comment (bojand) [reply rpadmin/admin.go:219]: Good callout. Added a mutex to protect the urls.

**Review by weeco — APPROVED** [review]

File: rpadmin/admin.go [hunk 8]
```diff
@@ -630,3 +675,67 @@ func MaxRetries(r int) Opt {
 		cl.MaxRetries = r
 	}}
 }
+
+// AdminAddressesFromK8SDNS attempts to deduce admin API URLs
+// based on Kubernetes DNS resolution.
+// We are in Kubernetes.
```
Comment (weeco) [inline rpadmin/admin.go:681]: NIT: Maybe a comment how we caller should check whether we are in K8s.
